crates/daphne/src/messages/request.rs

```diff
@@ -24,15 +25,79 @@ macro_rules! impl_req_body {
     };
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(any(test, feature = "test-utils"), derive(deepsize::DeepSizeOf))]
+pub struct AggregationJobRequestHash(Vec<u8>);
+
+impl AggregationJobRequestHash {
+    pub fn get(&self) -> &[u8] {
+        &self.0
+    }
+
+    fn hash(bytes: &[u8]) -> Self {
+        Self(
+            ring::digest::digest(&ring::digest::SHA256, bytes)
+                .as_ref()
+                .to_vec(),
+        )
+    }
+}
+
+pub struct HashedAggregationJobReq {
+    pub request: AggregationJobInitReq,
+    pub hash: AggregationJobRequestHash,
+}
+
+impl HashedAggregationJobReq {
+    #[cfg(any(test, feature = "test-utils"))]
+    pub fn from_aggregation_req(version: DapVersion, request: AggregationJobInitReq) -> Self {
+        let mut buf = Vec::new();
+        request.encode_with_param(&version, &mut buf).unwrap();
+        Self {
+            request,
+            hash: AggregationJobRequestHash::hash(&buf),
+        }
+    }
+}
+
+impl ParameterizedEncode<DapVersion> for HashedAggregationJobReq {
+    fn encode_with_param(
+        &self,
+        encoding_parameter: &DapVersion,
+        bytes: &mut Vec<u8>,
+    ) -> Result<(), prio::codec::CodecError> {
+        self.request.encode_with_param(encoding_parameter, bytes)
+    }
+}
+
+impl ParameterizedDecode<DapVersion> for HashedAggregationJobReq {
+    fn decode_with_param(
+        decoding_parameter: &DapVersion,
+        bytes: &mut std::io::Cursor<&[u8]>,
+    ) -> Result<Self, prio::codec::CodecError> {
+        let start = usize::try_from(bytes.position())
+            .map_err(|e| prio::codec::CodecError::Other(Box::new(e)))?;
+        let request = AggregationJobInitReq::decode_with_param(decoding_parameter, bytes)?;
+        let end = usize::try_from(bytes.position())
+            .map_err(|e| prio::codec::CodecError::Other(Box::new(e)))?;
+
+        Ok(Self {
+            request,
+            hash: AggregationJobRequestHash::hash(&bytes.get_ref()[start..end]),
+        })
+    }
+}
+
 impl_req_body! {
-//  body type             | id type
-//  --------------------- | ----------------
-    Report                | ()
-    AggregationJobInitReq | AggregationJobId
-    AggregateShareReq     | ()
-    CollectionReq         | CollectionJobId
-    CollectionPollReq     | CollectionJobId
-    ()                    | ()
+    //  body type           | id type
+    //  --------------------| ----------------
+    Report                  | ()
+    AggregationJobInitReq   | AggregationJobId
+    HashedAggregationJobReq | AggregationJobId
```

Comment:
If you take the previous suggestion, then also update this.

Reply:
What do you mean by update this? import from `roles::helper`?

Reply:
Yes